dilated-file-transfer.md

```diff
@@ -0,0 +1,499 @@
+In both cases, they are treated as "an offer" although a directory may consist of dozens or more individual files.
+
+XXX include compression in this revision, or allow that to be a future enhancement?
+
+
+## Version Negotiation
+
+There is an existing file-transfer protocol which does not use Dilation (called "classic" in this document).
+Clients supporting newer versions of file-transfer (i.e. the one in this document) SHOULD offer backwards compatibility.
+
+In the mailbox protocol, applications can indicate version information.
+The classic file-transfer protocol doesn't use this feature so the version information is empty.
+This new protocol will include a dict like:
+
+```json
+{
+    "transfer-v1": {
+        "mode": "{send|receive|connect}",
+        "features": [],
+    }
+}
+```
+
+The version of the protocol is the `"-v1"` tag in `"transfer-v1"`.
+A peer supporting newer versions may include `"transfer-v2"` or `"transfer-v3"` etc.
+There is currently only one version: `1`.
+Versions are considered as integers, so the version tag MUST always be the entire tail of the string, MUST start with `-v` and MUST end ONLY with digits (that are an integer version bigger than 0).
+
+When multiple versions are present, a peer decides which version to use by comparing the "list of versions" that they each support and selects the highest from the intersection of these.
+For example, if 3 versions existed, the two peers may present their version information like:
+
+```
+    peer A: "transfer-v1": {}, "transfer-v3": {}
+    peer B: "transfer-v1": {}, "transfer-v2": {}
+```
+
+Each peer makes a list of versions the other peer accepts: `A=[1, 3]` and `B=[1, 2]`.
+Taking the intersection of these yields the list `[1]` and the biggest number in that list is "1" so that is the version selected.
+When possible, peers SHOULD provide backwards compatibility.
+Note that you must declare each previous version supported (this allows support for any older version to be withdrawn by implementations).
+
+### `"transfer-v1"`:
+
+The `"mode"` key indicates the desired mode of the peer.
+It has one of three values:
+* `"send"`: the peer will only send files (similar to classic transfer protocol)
+* `"receive"`: the peer only receive files (the flip side of the above)
+* `"connect"`: the peer will send and receive zero or more files before closing the session
+
+If both peers indicate `"receive"` then nothing will ever happen so they both SHOULD end the session and disconnect.
+If both peers indicate `"send"` then they SHOULD also end the session (although whichever sends the first Offer will induce a protocol error in the other peer).
+If one peer indicates `"connect"` and the other indicates either `"send"` or `"receive"` then the peers can still interoperate and the `"connect"` side MUST continue (although it MAY indicate the peer's lack of one capability e.g. by disabling part of its UI).
```

Comment:
Does this mean the connect feature is optional to support? I think it maybe should, but I am not sure it is right now.

Reply:
Yes, you could write a client that supports _only_ receive mode for example (e.g. the existing `wormhole receive` command).

Reply:
Or, existing GUI clients could use `"send"` and `"receive"` mode "for now" as they upgrade to the new protocol and introduce different UX for `"connect"` ...?

Reply:
It is, yes. Connect does not make much sense for CLI applications, for example.